Typecheck objects passed to K_POLL_EVENT_INITIALIZER

arbrauns opened the pull request:
Too many times have I accidentally passed a `struct k_sem **` or similar. Thankfully modern C can help us here.

andyross commented on include/zephyr/toolchain/common.h:
```diff
@@ -175,6 +175,10 @@
 
+// Same as BUILD_ASSERT(), but evaluates to an expression of value 0. Useful in macros.
+#define BUILD_ASSERT_EXPR(EXPR, MSG...) \
+	(0 * sizeof(struct { BUILD_ASSERT(EXPR, MSG); int dummy; }))
```
As mentioned elsewhere, it's probably even more useful for this to take a parameter to which it will evaluate if it doesn't blow up.

andyross reviewed (commented):
Some notes.  Seems clever.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -5698,13 +5698,35 @@ struct k_poll_event {
+#define K_POLL_ASSERT_VALID_OBJECT(_event_type, _event_obj) BUILD_ASSERT_EXPR( \
+	(_event_type) == K_POLL_TYPE_IGNORE || (\
+		(_event_type) == K_POLL_TYPE_SIGNAL && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_poll_signal *)\
```
Heh, actually it looks like we already have it, see SAME_TYPE(): https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L233

(Which is actually a poor name for this, as it accepts promotions and automatic conversions too)

arbrauns replied:
Now uses `SAME_TYPE()` and guards against C++.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -5698,13 +5698,35 @@ struct k_poll_event {
 	.state = K_POLL_STATE_NOT_READY, \
 	.mode = _event_mode, \
-	.unused = 0, \
+	.unused = 0 + K_POLL_ASSERT_VALID_OBJECT(_event_type, _event_obj), \
```
GCC statement expressions[1] are the normal tool for that.  `({ STATEMENT_A; STATEMENT_B; })` executes A and then B, but then evaluates syntactically as the result of B.

[1] An extension we do allow, because it's so pervasively supported and incredibly useful.  See https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html

arbrauns replied:
Those aren't allowed outside of function scopes, which would be a weird restriction to impose.

npitre commented:
You are missing `K_POLL_TYPE_FIFO_DATA_AVAILABLE` and have
`K_POLL_TYPE_DATA_AVAILABLE` twice.

I think this can be made much simpler while fully leveraging the compiler's
type checking ability, and also be C++ compatible. Consider this:

```
diff --git a/include/zephyr/kernel.h b/include/zephyr/kernel.h
index 32c554c159d..f4fb7765933 100644
--- a/include/zephyr/kernel.h
+++ b/include/zephyr/kernel.h
@@ -5686,14 +5686,14 @@ struct k_poll_event {
 
 	/** per-type data */
 	union {
-		void *obj;
-		struct k_poll_signal *signal;
-		struct k_sem *sem;
-		struct k_fifo *fifo;
-		struct k_queue *queue;
-		struct k_msgq *msgq;
+		void *obj, *typed_K_POLL_TYPE_IGNORE;
+		struct k_poll_signal *signal, *typed_K_POLL_TYPE_SIGNAL;
+		struct k_sem *sem, *typed_K_POLL_TYPE_SEM_AVAILABLE;
+		struct k_fifo *fifo, *typed_K_POLL_TYPE_FIFO_DATA_AVAILABLE;
+		struct k_queue *queue, *typed_K_POLL_TYPE_DATA_AVAILABLE;
+		struct k_msgq *msgq, *typed_K_POLL_TYPE_MSGQ_DATA_AVAILABLE;
 #ifdef CONFIG_PIPES
-		struct k_pipe *pipe;
+		struct k_pipe *pipe, *typed_K_POLL_TYPE_PIPE_DATA_AVAILABLE;
 #endif
 	};
 };
@@ -5706,7 +5706,7 @@ struct k_poll_event {
 	.mode = _event_mode, \
 	.unused = 0, \
 	{ \
-		.obj = _event_obj, \
+		.typed_##_event_type = _event_obj, \
 	}, \
 	}
 
@@ -5719,7 +5719,7 @@ struct k_poll_event {
 	.mode = _event_mode, \
 	.unused = 0, \
 	{ \
-		.obj = _event_obj, \
+		.typed_##_event_type = _event_obj, \
 	}, \
 	}
 
```

This even trips an existing test that does this:

```
        struct k_poll_event bad_events2[] = {
                K_POLL_EVENT_INITIALIZER(0xFU,
                                         K_POLL_MODE_NOTIFY_ONLY,
                                         &no_wait_sem),
        };
```

And I think that's a good thing. Initializers shouldn't be used with random
literals. The test could simply be modified to work around the initializer
in this particular case.

cfriedt commented:
Sort of leaning toward Nico's solution and allowing the compiler to better do its job.

arbrauns committed:
include/kernel: check type of object passed to K_POLL_EVENT_INITIALIZER

This prevents accidentally passing the wrong pointer type (e.g. one level
of indirection too many) as the object `void *`.

Co-authored-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns commented:
> You are missing K_POLL_TYPE_FIFO_DATA_AVAILABLE and have
K_POLL_TYPE_DATA_AVAILABLE twice.

Those are the same thing.

Anyway, I've adopted your solution, it is nicer indeed.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -5686,14 +5686,14 @@ struct k_poll_event {
+		struct k_sem *sem, *typed_K_POLL_TYPE_SEM_AVAILABLE;
+		struct k_fifo *fifo, *typed_K_POLL_TYPE_FIFO_DATA_AVAILABLE;
+		struct k_queue *queue, *typed_K_POLL_TYPE_DATA_AVAILABLE;
+		struct k_msgq *msgq, *typed_K_POLL_TYPE_MSGQ_DATA_AVAILABLE;
```
Might not be the worst thing to have a comment here to the effect of  " the 'typed_' fields exist for the benefit of the initializers below that take the type symbols as an argument" or whatever.

andyross reviewed (dismissed):
Very clean

kartben commented:
@andyross need you to re-approve (I know you love this :)) thanks!

andyross reviewed (approved):
Fell behind on reviews, but still +1

evgeniy-paltsev commented:
This breaks existing Zephyr tests.


For example`tests/kernel/poll`:
How to reproduce:

```
west build -b qemu_x86 tests/kernel/poll -T kernel.poll
```

```
FAILED: CMakeFiles/app.dir/src/test_poll.c.obj 
In file included from /zephyr/subsys/testsuite/ztest/include/zephyr/ztest.h:40,
                 from /zephyr/tests/kernel/poll/src/test_poll.c:7:
/zephyr/tests/kernel/poll/src/test_poll.c: In function 'poll_api_1cpu_test_poll_no_wait':
/zephyr/include/zephyr/kernel.h:5757:18: error: 'union <anonymous>' has no member named 'typed_0xFU'
 5757 |                 .typed_##_event_type = _event_obj, \
      |                  ^~~~~~
/zephyr/tests/kernel/poll/src/test_poll.c:119:17: note: in expansion of macro 'K_POLL_EVENT_INITIALIZER'
  119 |                 K_POLL_EVENT_INITIALIZER(0xFU,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
[23/144] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/cbprintf_packaged.c.obj
```

@jhedberg can we revert it?

npitre commented:
Better yet, let's fix the test instead.
The test is definitively wrong.

npitre commented:
Failing test fixed in PR #74126.
